File: runtime/python/prompty/prompty/core/context.py
"""Context window management — trimming and summarization.

When agent loop conversations grow too long, this module trims older
non-system messages and produces a compact summary to preserve context.
"""

from __future__ import annotations

from .types import Message, TextPart

__all__ = [
    "estimate_chars",
    "format_dropped_messages",
    "summarize_dropped",
    "trim_to_context_window",
]


def format_dropped_messages(messages: list[Message]) -> str:
    """Format dropped messages as readable text for compaction prompts.

    Each message is rendered as ``[role]: text`` with tool calls shown
    as ``Called: name(args)``.
    """
    import json

    lines: list[str] = []
    for msg in messages:
        text = msg.text.strip() if msg.text else ""
        if text:
            lines.append(f"[{msg.role}]: {text}")
        tool_calls = msg.metadata.get("tool_calls")
        if tool_calls and isinstance(tool_calls, list):
            for tc in tool_calls:
                name = tc.get("name", tc.get("function", {}).get("name", "?"))
                args = tc.get("arguments", tc.get("function", {}).get("arguments", ""))
                if isinstance(args, dict):
                    args = json.dumps(args)
                lines.append(f"Called: {name}({args})")
    return "\n".join(lines)


def estimate_chars(messages: list[Message]) -> int:
    """Estimate the character cost of a message list.

    Per spec §13.3: role + 4 overhead per message, text parts by length,
    non-text parts at a fixed 200-char estimate, tool_calls by JSON length.
    """
    import json

    total = 0
    for msg in messages:
        total += len(msg.role) + 4
        for part in msg.parts:
            if isinstance(part, TextPart):
                total += len(part.value)
            else:
                total += 200
        tool_calls = msg.metadata.get("tool_calls")
        if tool_calls:
            total += len(json.dumps(tool_calls, default=str))
    return total


def _truncate(text: str, max_len: int = 200) -> str:
    """Truncate a string to at most *max_len* characters."""
    if len(text) <= max_len:
        return text
    return text[:max_len] + "…"


def summarize_dropped(messages: list[Message]) -> str:
    """Build a compact string summary from dropped messages.

    Extracts user requests, assistant decisions, and tool actions
    without needing an LLM call.
    """
    lines: list[str] = []
    for msg in messages:
        text = msg.text.strip() if msg.text else ""
        if msg.role == "user":
            if text:
                lines.append(f"User asked: {_truncate(text)}")
        elif msg.role == "assistant":
            if text:
                lines.append(f"Assistant: {_truncate(text)}")
            tool_calls = msg.metadata.get("tool_calls")
            if tool_calls and isinstance(tool_calls, list):
                names = [tc.get("name", tc.get("function", {}).get("name", "?")) for tc in tool_calls]
                lines.append(f"  Called tools: {', '.join(names)}")
        # Skip tool-result messages (captured in assistant summary)

    if not lines:
        return ""

    # Cap the summary at ~4000 chars
    result = "[Context summary: "
    for line in lines:
        if len(result) + len(line) > 4000:
            result += "\n... (older messages omitted)"
            break
        result += line + "\n"
    result = result.rstrip() + "]"
    return result


def trim_to_context_window(
    messages: list[Message],
    budget_chars: int,
) -> tuple[int, list[Message]]:
    """Trim messages to fit within a character budget.

    Strategy per spec §13.3:
    1. Keep system messages at the front
    2. Reserve space for summary (~5000 chars or 5% of budget)
    3. Drop oldest non-system messages until within budget
    4. Summarize dropped messages and insert after system messages

    Returns
    -------
    tuple[int, list[Message]]
        (count of dropped messages, the dropped messages themselves)
    """
    if estimate_chars(messages) <= budget_chars:
        return 0, []

    # Partition: leading system messages vs rest
    system_end = 0
    for i, msg in enumerate(messages):
        if msg.role != "system":
            system_end = i
            break
    else:
        system_end = len(messages)

    system_msgs = messages[:system_end]
    rest = messages[system_end:]

    summary_budget = min(5000, int(budget_chars * 0.05))

    # Drop oldest non-system messages
    dropped: list[Message] = []
    while estimate_chars(system_msgs + rest) > (budget_chars - summary_budget) and len(rest) > 2:
        dropped.append(rest.pop(0))

    dropped_count = len(dropped)

    if dropped_count > 0:
        summary_text = summarize_dropped(dropped)
        if summary_text:
            summary_msg = Message(
                role="user",
                parts=[TextPart(value=summary_text)],
            )
            messages.clear()
            messages.extend(system_msgs)
            messages.append(summary_msg)
            messages.extend(rest)
        else:
            messages.clear()
            messages.extend(system_msgs)
            messages.extend(rest)
    else:
        messages.clear()
        messages.extend(system_msgs)
        messages.extend(rest)

    return dropped_count, dropped

File: runtime/typescript/packages/core/src/core/context.ts
/**
 * §13.3 Context Window Management — trimming and summarization.
 * @module
 */

import { Message, TextPart } from "./types.js";

/**
 * Estimate the character cost of a message list.
 * Per spec §13.3: role + 4 overhead per message, text parts by length,
 * non-text parts at 200-char estimate.
 */
export function estimateChars(messages: Message[]): number {
  let total = 0;
  for (const msg of messages) {
    total += msg.role.length + 4;
    for (const part of msg.parts) {
      if (part.kind === "text") {
        total += (part as TextPart).value.length;
      } else {
        total += 200;
      }
    }
    const toolCalls = msg.metadata?.tool_calls;
    if (toolCalls) {
      total += JSON.stringify(toolCalls).length;
    }
  }
  return total;
}

function truncate(text: string, maxLen = 200): string {
  return text.length <= maxLen ? text : text.slice(0, maxLen) + "…";
}

/**
 * Build a compact string summary from dropped messages.
 */
export function summarizeDropped(messages: Message[]): string {
  const lines: string[] = [];
  for (const msg of messages) {
    const msgText = msg.text.trim();
    if (msg.role === "user" && msgText) {
      lines.push(`User asked: ${truncate(msgText)}`);
    } else if (msg.role === "assistant") {
      if (msgText) lines.push(`Assistant: ${truncate(msgText)}`);
      const toolCalls = msg.metadata?.tool_calls;
      if (Array.isArray(toolCalls)) {
        const names = toolCalls.map(
          (tc: Record<string, unknown>) =>
            (tc.name as string) ?? ((tc.function as Record<string, string>)?.name ?? "?"),
        );
        lines.push(`  Called tools: ${names.join(", ")}`);
      }
    }
  }
  if (lines.length === 0) return "";

  let result = "[Context summary: ";
  for (const line of lines) {
    if (result.length + line.length > 4000) {
      result += "\n... (older messages omitted)";
      break;
    }
    result += line + "\n";
  }
  return result.trimEnd() + "]";
}

/**
 * Format dropped messages as a human-readable text block.
 * Each message is rendered as `[role]: content`, with tool calls shown as
 * `Called: name(args)`.
 */
export function formatDroppedMessages(messages: Message[]): string {
  const lines: string[] = [];
  for (const msg of messages) {
    const msgText = msg.text.trim();
    if (msgText) {
      lines.push(`[${msg.role}]: ${msgText}`);
    }
    const toolCalls = msg.metadata?.tool_calls;
    if (Array.isArray(toolCalls)) {
      for (const tc of toolCalls) {
        const tcObj = tc as Record<string, unknown>;
        const name = (tcObj.name as string) ?? ((tcObj.function as Record<string, string>)?.name ?? "?");
        const args = (tcObj.arguments as string) ?? ((tcObj.function as Record<string, string>)?.arguments ?? "");
        lines.push(`Called: ${name}(${args})`);
      }
    }
  }
  return lines.join("\n");
}

/**
 * Trim messages in-place to fit within a character budget.
 * Returns [droppedCount, droppedMessages].
 */
export function trimToContextWindow(
  messages: Message[],
  budgetChars: number,
): [number, Message[]] {
  if (estimateChars(messages) <= budgetChars) {
    return [0, []];
  }

  // Partition: leading system messages vs rest
  let systemEnd = 0;
  for (let i = 0; i < messages.length; i++) {
    if (messages[i].role !== "system") {
      systemEnd = i;
      break;
    }
    if (i === messages.length - 1) systemEnd = messages.length;
  }

  const systemMsgs = messages.slice(0, systemEnd);
  const rest = messages.slice(systemEnd);
  const summaryBudget = Math.min(5000, Math.floor(budgetChars * 0.05));
  const dropped: Message[] = [];

  while (estimateChars([...systemMsgs, ...rest]) > budgetChars - summaryBudget && rest.length > 2) {
    dropped.push(rest.shift()!);
  }

  const droppedCount = dropped.length;

  // Rebuild messages array in-place
  messages.length = 0;
  messages.push(...systemMsgs);

  if (droppedCount > 0) {
    const summaryText = summarizeDropped(dropped);
    if (summaryText) {
      messages.push(new Message({ role: "user", parts: [new TextPart({ value: summaryText })] }));
    }
  }

  messages.push(...rest);
  return [droppedCount, dropped];
}
